### PR TITLE
feature: get unitary for an entire circuit

### PR DESCRIPTION
**Pull request by indisoluble**
*Issue #, if available:* https://github.com/aws/amazon-braket-sdk-python/issues/209

*Description of changes:* New simulator to produce the unitary matrix that represents an entire circuit

*Testing done:* Enough to cover 100% of the code, however more might be required, please let me know

**Notice:** These changes are pointing to another PR https://github.com/aws/amazon-braket-schemas-python/pull/92 where I added a new result named `UnitaryMatrix`. This PR has to be merged before this one and, when it is done, I will update this one to point again to AWS `main` branch

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/default_simulator/unitary_matrix_simulation.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    ) -> np.ndarray:
+        subscripts = UnitaryMatrixSimulation._einsum_subscripts(targets, qubit_count)
+
+        target_count = len(targets)
+        gate_matrix = np.reshape(matrix, target_count * [2, 2])
+
+        return opt_einsum.contract(subscripts, gate_matrix, un_tensor, dtype=complex, casting="no")
+
+    @staticmethod
+    def _einsum_subscripts(targets: Tuple[int, ...], qubit_count: int) -> str:
```
Comment (indisoluble) [inline src/braket/default_simulator/unitary_matrix_simulation.py:83]: This code is based in this one in Qiskit Terra: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/providers/basicaer/basicaertools.py#L132

File: test/unit_tests/braket/default_simulator/test_unitary_matrix_simulation.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import numpy as np
+import pytest
+
+from braket.default_simulator import gate_operations
+from braket.default_simulator.unitary_matrix_simulation import UnitaryMatrixSimulation
+
+evolve_testdata = [
```
Comment (indisoluble) [inline test/unit_tests/braket/default_simulator/test_unitary_matrix_simulation.py:20]: I wrote these tests based on the tests for the unitary simulator in Qiskit Terra: https://github.com/Qiskit/qiskit-terra/blob/main/test/python/basicaer/test_unitary_simulator.py

**Review by shiyunon — COMMENTED** [review]
Several questions:

1. I don't see the code throw any error messages if the circuit has components that have nounitary matrix definitions. These components include: 1. quantum noises 2. (in the future) compiler options like `barrier` or classical operations 3. (in the future) parametric gates. Were these corner cases considered?
2. If you implemented a unitary simulation, why encode UnitaryMatrix as a Result type but not just let users do `circ.unitary()`?
3. It's not clear to me if add another "simulator" is necessary if we already have LocalSimulator and SV1.

**Review by shiyunon — COMMENTED** [review]
Thanks for the PR. 

I have several questions before a full review from the team:

1. I don't see the code throw any error messages if the circuit has components that don't have unitary matrix definitions. These components include: 1. quantum noises 2.  compiler options like `barrier` or classical operations (this is future-proof, Braket doesn't have it yet) 3. parametric gates  (this is also future-proof, Braket doesn't have it yet).
2. If you implemented a unitary simulation, why encode UnitaryMatrix as a result type but not just let users do `circ.unitary()`?
3. It's not clear to me if add another "simulator" is necessary if we already have LocalSimulator and SV1.

**Comment by indisoluble** [comment]
Hi @shiyunon, thanks for your comments :) Let's go one by one:

1. In order to implement `UnitaryMatrixSimulation.evolve()`, I followed the example of `StateVectorSimulation.evolve()` and `DensityMatrixSimulation.evolve()`; in both cases, although I might be wrong about this, they get the `matrix` & `targets` from a `GateOperation` instance and then perform the necessary operations. As I see it, it is on  `GateOperation` to validate the information returned but this might be a wrong assumption, if there is an example somewhere else in the code that I can follow, please let me know and thanks on advance :+1: 
2. Well, given that I coded a new simulator, I do need a new result type, otherwise [BaseLocalSimulator._validate_ir_results_compatibility()](https://github.com/indisoluble/amazon-braket-default-simulator-python/blob/circuit_unitary/src/braket/default_simulator/simulator.py#L120) will always fail, right?... although your next question is: why did I code a new simulator? Anyway, if by `circ` you mean the `braket.circuits.Circuit`, I have not modified [amazon-braket-sdk-python](https://github.com/aws/amazon-braket-sdk-python) ... yet :)
3. `braket.devices.LocalSimulator` is part of [amazon-braket-sdk-python](https://github.com/aws/amazon-braket-sdk-python). I have modified [amazon-braket-default-simulator-python](https://github.com/aws/amazon-braket-default-simulator-python) to provide a new device that `braket.devices.LocalSimulator` can use to extend its capabilities.

Regards.

**Comment by shiyunon** [comment]
> Hi @shiyunon, thanks for your comments :) Let's go one by one:
> 
> 1. In order to implement `UnitaryMatrixSimulation.evolve()`, I followed the example of `StateVectorSimulation.evolve()` and `DensityMatrixSimulation.evolve()`; in both cases, although I might be wrong about this, they get the `matrix` & `targets` from a `GateOperation` instance and then perform the necessary operations. As I see it, it is on  `GateOperation` to validate the information returned but this might be a wrong assumption, if there is an example somewhere else in the code that I can follow, please let me know and thanks on advance 👍
> 2. Well, given that I coded a new simulator, I do need a new result type, otherwise [BaseLocalSimulator._validate_ir_results_compatibility()](https://github.com/indisoluble/amazon-braket-default-simulator-python/blob/circuit_unitary/src/braket/default_simulator/simulator.py#L120) will always fail, right?... although your next question is: why did I code a new simulator? Anyway, if by `circ` you mean the `braket.circuits.Circuit`, I have not modified [amazon-braket-sdk-python](https://github.com/aws/amazon-braket-sdk-python) ... yet :)
> 3. `braket.devices.LocalSimulator` is part of [amazon-braket-sdk-python](https://github.com/aws/amazon-braket-sdk-python). I have modified [amazon-braket-default-simulator-python](https://github.com/aws/amazon-braket-default-simulator-python) to provide a new device that `braket.devices.LocalSimulator` can use to extend its capabilities.
> 
> Regards.

1. Got it, I saw your supported gate list. Yes, it is indeed the case, simulators just validate circuits like QPUs, so there shouldn't be problems with it.
2 &3. In that case, I think it will be a better user experience if we can have this in the braket_sdk (at least for small circuits), but these design problems will probably have to be answered after the team's full review.

Thanks again for the PR!

**Comment by indisoluble** [comment]
By the way, I checked the "dependent tests" that actually failed (like this one: https://github.com/aws/amazon-braket-default-simulator-python/pull/71/checks?check_run_id=2723103546#step:5:2973) and it seems like it is not able to find `braket.ir.jaqcd.UnitaryMatrix` but I did this change ... in my fork for [amazon-braket-schemas-python](https://github.com/indisoluble/amazon-braket-schemas-python). I guess the test got the `main` branch for AWS `amazon-braket-schemas-python` instead of mine. Hopefully, tests will pass once (and if) https://github.com/aws/amazon-braket-schemas-python/pull/92 is merged.

File: src/braket/default_simulator/unitary_matrix_simulation.py (hunk 1 above)
Comment (nonhermitian) [reply src/braket/default_simulator/unitary_matrix_simulation.py:83]: If using Qiskit code here, it would be nice to acknowledge that fact.

Comment (indisoluble) [reply src/braket/default_simulator/unitary_matrix_simulation.py:83]: Sure :) Although this code is not a copy of that one in Qiskit so, as I see it, it does not feel under the requirements of Apache license ... but I have to admit I am not sure I fully understand all the details of Apache license, I might be wrong about this. Nonetheless, it is worth to reference the sources, should I add a comment in the code? Some advice on this regard would be appreciated.

File: test/unit_tests/braket/default_simulator/test_unitary_matrix_simulation.py (hunk 2 above)
Comment (nonhermitian) [reply test/unit_tests/braket/default_simulator/test_unitary_matrix_simulation.py:20]: This should have a copyright statement of Qiskit somewhere in the file as this is obvious a derivative work of the original.

File: src/braket/default_simulator/unitary_matrix_simulation.py (hunk 1 above)
Comment (nonhermitian) [reply src/braket/default_simulator/unitary_matrix_simulation.py:83]: So, per Apache 2, "derivative works" must contain the copyright as well.  However there is of course gray area.  Perhaps here just cite the original source of the code in question.  However, it is really up to the maintainers of this repo to decide such things.

Comment (indisoluble) [reply src/braket/default_simulator/unitary_matrix_simulation.py:83]: Thank you @nonhermitian for your explanation!!! I will wait then for a comment from the reviewers. Regards

**Comment by kshitijc** [comment]
Thank you for the PR @indisoluble! We respect the open source policies and would like to have attribution notices where necessary. 
Please allow some time for us to review this request, as this is an important feature and requires some collaborative decisions on the team.

**Comment by kshitijc** [comment]
Thanks for your patience @indisoluble! After a careful review, we think that the unitary matrix representation functionality would be better suited to live in the amazon-braket-sdk-python repository, and be accessible to customers through a method on the Circuit class (`as_unitary()`) allowing the users to get the unitary representation of the circuit easily.  Here are some considerations which guided us towards this decision:
* Placing this functionality in a simulator would involve much more work, both from the users (the unitary representation of a Circuit now needs to be accessed in a roundabout way through a simulator) and from the maintainers of this code (there is additional code that goes with a device, which needs to be maintained).
* This requires a UnitaryMatrix result type, which can be a source of confusion for some of the customers (who are directly interacting with the service) since no other devices support it.
* There’s performance overhead in translating from the circuit class to the representation expected by the simulator, which can be avoided.

**Comment by shiyunon** [comment]
Hi, @indisoluble. I complement @kshitijc 's review with the following points:

As for the Qiskit code that was used, in general, we have no problem with reusing code from another framework as long as we abide by the requirements of the Apache licence. However, it indeed requires more considerations for some of the choices here. On line 80 in `unitary_matrix_simulation.py`, the `contract` method from the`opt_einsum` package was used. We find it might be worthwhile to compare betewen the `opt_einsum` package and default Numpy `einsum`. Here are the tradeoffs:
1. `opt_einsum` is much faster for tensor contractions with many contract paths to choose from. For this simple usage, we expect the performance difference to be minimal for this use case but it requires further benchmarking.
2.  Using`opt_einsum` adds 3rd party dependence to Braket which increases complexity in future maintenance.
 
With the above said, it is beyond the scope of this PR to make the decision. We will respect your choice here and will further investigate into it on the team (so you don't have to do anything). The changes we do recommend you to make are:
1. For the `_einsum_subscripts` defined on line 83 in `unitary_matrix_simulation.py`, we recommend to implement this method using the `np.array` calling style instead of the `str` calling style to improve the readability. Specifically, both `einsum` and `opt_einsum` allow you to call `einsum` (or `contract` in this case) in two ways: (1) supply the subscripts as strings (like what you did here) (2) supply the subscripts as numpy.array. An example would be:
```
>>> a
array([[1, 2],
       [3, 4]])
>>> opt_einsum.contract('ij->ji', a)
array([[1, 3],
       [2, 4]])
>>> opt_einsum.contract(a, [0, 1], [1, 0])
array([[1, 3],
       [2, 4]])
```
We recommend the second way of supplying the subscripts to increase readability.

2. For line 85-85, in `unitary_matrix_simulation.py`, with the `numpy.array` way of supplying the subscripts, there shouldn't be a limit of 26 for the indices anymore. However, for user visibility, we do want the function to warn the users that `as_unitary()` can get slow if the qubit count is beyond 10 qubits. Please add the proper warning to this function.
3. Please add more test cases in test file `test_unitary_matrix_simulation.py`  to have a better test coverage. For example, the current test doesn’t test all gates. 

If you would make the above changes, we will review the new commit to see if an acknowledgement of Qiskit is needed. 

Again, thanks for your contribution here to Braket!

**Comment by indisoluble** [comment]
Hi @kshitijc & @shiyunon,

Thanks for your comments, I will apply them (hopefully soon) in a new PR for  `amazon-braket-sdk-python`. Right now, I will close this PR as it is no longer necessary.

Regards.

**Pull request closed without merging**